File: src/main/java/ir/ut/ai/search/BFSearch.java
package ir.ut.ai.search;

import ir.ut.ai.base.Action;
import ir.ut.ai.base.Node;
import ir.ut.ai.base.Problem;

import java.util.ArrayList;
import java.util.List;

public class BFSearch<S, A> extends SearchBase<S, A> {

    private List<Node<S, A>> frontier;
    private List<Node<S, A>> explored;

    public BFSearch(Problem<S, A> problem) {
        super(problem);
        this.frontier = new ArrayList<>();
        this.explored = new ArrayList<>();
    }

    @Override
    protected List<Node<S, A>> solution(Node<S, A> node) {
        return explored;
    }

    @Override
    public List<Node<S, A>> run() {
        Node node = problem.getInitialNode();
        if (problem.isGoal(node)) {
            return new ArrayList<Node<S, A>>() {{
                add(node);
            }};
        }
        frontier.add(node);
        while (!frontier.isEmpty()) {
            Node cNode = frontier.remove(0);
            explored.add(cNode);
            if (problem.isGoal(cNode)) {
                return solution(cNode);
            }
            ArrayList<Action<A>> actionList = problem.getActions().get(cNode);
            if (actionList == null)
                actionList = new ArrayList<>();
            for (Action action : actionList) {
                Node nextNode = nexNode(cNode, action);
                if (nextNode != null) {
                    if (!frontier.contains(nextNode) || !explored.contains(nextNode)) {
                        frontier.add(nextNode);
                    }
                }
            }
        }
        return new ArrayList<>();
    }
}

File: src/main/java/ir/ut/ai/search/BFSearch.kt
package ir.ut.ai.search

import ir.ut.ai.base.Action
import ir.ut.ai.base.Node
import ir.ut.ai.base.Problem

import java.util.ArrayList

class BFSearch<S, A>(problem: Problem<S, A>) : SearchBase<S, A>(problem) {

    private val frontier: MutableList<Node<S, A>> = ArrayList<Node<S, A>>()
    private val explored: MutableList<Node<S, A>> = ArrayList<Node<S, A>>()

    override fun solution(node: Node<S, A>): List<Node<S, A>> {
        return explored
    }

    override fun run(): List<Node<S, A>> {
        val node = problem.initialNode
        if (problem.isGoal(node)) {
            return object : ArrayList<Node<S, A>>() {
                init {
                    add(node)
                }
            }
        }
        frontier.add(node)
        while (!frontier.isEmpty()) {
            val cNode = frontier.removeAt(0)
            explored.add(cNode)
            if (problem.isGoal(cNode)) {
                return solution(cNode)
            }
            var actionList: ArrayList<Action<A>>? = problem.actions[cNode]
            for (action in actionList!!) {
                val nextNode = nexNode(cNode, action)
                if (nextNode != null) {
                    if (!frontier.contains(nextNode) || !explored.contains(nextNode)) {
                        frontier.add(nextNode)
                    }
                }
            }
        }
        return ArrayList()
    }
}
